elasticsearch/src/main/java/com/gentics/mesh/search/verticle/bulk/BulkOperator.java

```diff
@@ -85,12 +89,22 @@ public void drain() {
 					if (!canceled.get() && requested.get() > 0 && !bulkableRequests.isEmpty() && flushing.compareAndSet(true, false)) {
 						timer.stop();
 						log.trace("Emitting bulk of size {} to subscriber", bulkableRequests.size());
-						BulkRequest request = new BulkRequest(bulkableRequests.asList());
-						bulkableRequests.clear();
-						if (log.isDebugEnabled()) {
-							log.debug("Sending bulk to elasticsearch:\n{}", request);
-						}
-						subscriber.onNext(request);
+						do {
+							Long[] bulkLength = new Long[] { 0L };
+							List<Bulkable> requests = IntStream.range(0, requestLimit)
+								.mapToObj(index -> bulkableRequests.poll())
+								.takeWhile(maybeBulkable -> maybeBulkable
+										.filter(bulkable -> bulkLength[0] < lengthLimit)
+										.isPresent())
+								.map(Optional::get)
+								.peek(bulkable -> bulkLength[0] += bulkable.bulkLength())
+								.collect(Collectors.toList());
+							BulkRequest request = new BulkRequest(requests);
+							if (log.isDebugEnabled()) {
+								log.debug("Sending bulk to elasticsearch:\n{}", request);
+							}
+							subscriber.onNext(request);
+						} while (!bulkableRequests.isEmpty());
 						BackpressureHelper.produced(requested, 1);
```

Comment:
Doesn't this have to be changed as well, or at least moved to the do-while loop?